smolyay/samples.py

```diff
@@ -508,3 +511,583 @@ def _create(self):
+        p_gen = scipy.stats.qmc.LatinHypercube(
+            self.num_dimensions,
+            scramble=self.scramble,
+            strength=self.strength,
+            optimization=self.optimization,
+            seed=self.seed,
+        ).random(n=self.num_points)
+        self._points = scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class HaltonRandomPointSet(_QMCRandomPointSet):
+    """Generates a grid using Halton Sequences
+
+    Parameters
+    ----------
+    domain: list
+        domain of the random points
+
+    num_points: int
+        number of points to generate
+
+    seed: int
+        seed for generating the random points
+
+    scramble : bool
+        Default True. Applies centering to Latin Hypercube points, Owen
+        scrambling to Halton points, and LMS+shift scrambling to Sobol points.
+
+    optimization : {None, "random-cd", "lloyd"}
+        Default None. If "random-cd" the coordinates of points are adjusted to
+        lower the centered discrepancy. If "lloyd", adjust points using a
+        Lloyd-Max algorithm to encourage even spacing.
+
+    bits : int
+        Default 30. Sets the max number of points that can be generated,
+        which is 2**bits.
+    """
+
+    def _create(self):
+        """Generate a set of random points
+
+        Generates a set of quasi-monte carlo Halton Sequence points.
+        """
+        lower_bounds = [bound[0] for bound in self.domain]
+        upper_bounds = [bound[1] for bound in self.domain]
+        p_gen = scipy.stats.qmc.Halton(
+            self.num_dimensions,
+            scramble=self.scramble,
+            optimization=self.optimization,
+            seed=self.seed,
+        ).random(n=self.num_points)
+        self._points = scipy.stats.qmc.scale(p_gen, lower_bounds, upper_bounds)
+
+
+class SobolRandomPointSet(_QMCRandomPointSet):
+    """Generates a grid using Sobol Sequence
+
+    Parameters
+    ----------
+    domain: list
+        domain of the random points
+
+    num_points: int
+        number of points to generate. Must be a power of 2.
+
+    seed: int
+        seed for generating the random points
+
+    scramble : bool
+        Default True. Applies centering to Latin Hypercube points, Owen
+        scrambling to Halton points, and LMS+shift scrambling to Sobol points.
+
+    optimization : {None, "random-cd", "lloyd"}
+        Default None. If "random-cd" the coordinates of points are adjusted to
+        lower the centered discrepancy. If "lloyd", adjust points using a
+        Lloyd-Max algorithm to encourage even spacing.
+
+    Raises
+    ------
+    ValueError
+        number of points must be a power of two
+    ValueError
+        number of points must be less than 2**bits
+    """
+
+    def __init__(
+        self, domain, num_points, seed, scramble=True, optimization=None, bits=30
+    ):
+        self._bits = 64  # max value of bits
+        super().__init__(domain, num_points, seed, scramble, optimization)
```

Comment:
This is an anti-pattern here. Call super first unless there is a reason not to, and then I prefer `None` if you are going to overwrite immediately anyway.
```suggestion
        super().__init__(domain, num_points, seed, scramble, optimization)
        self._bits = None
```

Reply:
The super calls the number of points setter which checks if the number of points exceeds bits. This means that bits needs to be set to something before the number of points setter is called, so here it's set to the max to ensure that the "Number of points must be less than 2**bits" error doesn't occur.

Reply:
I see, please add a more explicit comment about why this is coming in this order then

Reply:
I did. It should be ready for review again now.